### PR TITLE
harvest fees when closing token2022 token account

### PR DESCRIPTION
**Pull request by urieltan**

### CODE REVIEW

File: programs/locker/src/util/token2022.rs [hunk 1]
```diff
@@ -219,6 +222,40 @@ pub fn calculate_transfer_fee_included_amount(
...
+        return Result::Ok(());
+    }
+
+    let token_mint_data = mint_info.try_borrow_data()?;
+    let token_mint_unpacked =
+        StateWithExtensions::<spl_token_2022::state::Mint>::unpack(&token_mint_data)?;
+    if let Ok(_transfer_fee_config) = token_mint_unpacked.get_extension::<TransferFeeConfig>() {
+        let escrow_state = escrow.load()?;
+        let escrow_seeds = escrow_seeds!(escrow_state);
+        harvest_withheld_tokens_to_mint(
```
Comment (andrewsource147) [inline programs/locker/src/util/token2022.rs:243]: Should check withheld_amount is non-zero before doing this CPI

File: programs/locker/src/util/token2022.rs [hunk 2]
```diff
@@ -219,6 +222,40 @@ pub fn calculate_transfer_fee_included_amount(
...
+            CpiContext::new_with_signer(
+                token_program_id.to_account_info(),
+                HarvestWithheldTokensToMint {
+                    token_program_id: token_program_id.to_account_info(),
+                    mint: mint.to_account_info(),
+                },
+                &[&escrow_seeds[..]],
```
[the remark this answers is not in the document]
Comment (urieltan) [reply programs/locker/src/util/token2022.rs:250]: What do you mean? the CPI doesn't need a signer?

[the remark this answers is not in the document]
Comment (andrewsource147) [reply programs/locker/src/util/token2022.rs:250]: the endpoint hasvest fee is permissionless, so we dont need signer

File: programs/locker/src/util/token2022.rs [hunk 3]
```diff
@@ -219,6 +222,43 @@ pub fn calculate_transfer_fee_included_amount(
     Ok(actual_amount)
 }
 
+// This function calculate the pre amount (with fee) require to transfer `amount` of token
```
Comment (andrewsource147) [inline programs/locker/src/util/token2022.rs:225]: should fix the comment here

File: programs/locker/src/util/token2022.rs [hunk 4]
```diff
@@ -219,6 +222,43 @@ pub fn calculate_transfer_fee_included_amount(
...
+pub fn harvest_fees_if_available<'c: 'info, 'info>(
```
Comment (andrewsource147) [inline programs/locker/src/util/token2022.rs:226]: should rename the function to `harvest_fees` because we dont check withheld_amount != 0

**Review by andrewsource147 — APPROVED** [review]

**Pull request merged**